Add Prophet pyfunc inline example (#7719)
mlflow/mlflow: docs/source/models.rst

You can also use the :py:func:`mlflow.prophet.load_model()`
method to load MLflow Models with the ``prophet`` model flavor in native prophet format.

Prophet pyfunc usage
~~~~~~~~~~~~~~~~~~~~

This example uses the `example_wp_log_peyton_manning.csv <https://github.com/facebook/prophet/blob/main/examples/example_wp_log_peyton_manning.csv>` _
from Prophet's GitHub repository. A sample of the dataset is as follows:

============= =================
ds            y
============= =================
2007-12-10    9.59076113897809
2007-12-11    8.51959031601596
2007-12-12    8.18367658262066
2007-12-13    8.07246736935477
============= =================

.. code-block:: py

    import pandas as pd
    from prophet import Prophet
    import mlflow

    train_df = pd.read_csv("https://raw.githubusercontent.com/facebook/prophet/main/examples/example_wp_log_peyton_manning.csv")

>     reg_model = Prophet(changepoint_prior_scale=0.5, uncertainty_samples=7)

Reviewer: what does 'reg' stand for here?  Can we just name the variable `prophet_model` ?

Author: "reg" is short for "regression" ;) Anyway, renamed as suggested.

You can also use the :py:func:`mlflow.prophet.load_model()`
method to load MLflow Models with the ``prophet`` model flavor in native prophet format.

Prophet pyfunc usage
~~~~~~~~~~~~~~~~~~~~

> This example uses the `example_wp_log_peyton_manning.csv <https://github.com/facebook/prophet/blob/main/examples/example_wp_log_peyton_manning.csv>` _

Reviewer: instead of displaying the name of the dataset, perhaps use the hyperlink key to explain what the dataset is.

Author: Done.

> from Prophet's GitHub repository. The dataset looks like this:

Reviewer: ```suggestion
from Prophet's GitHub repository. A sample of the dataset is as follows:
```

Author: Done.

============= =================
ds            y
============= =================
2007-12-10    9.59076113897809
2007-12-11    8.51959031601596
2007-12-12    8.18367658262066
2007-12-13    8.07246736935477
============= =================

.. code-block:: py

    import pandas as pd
    from prophet import Prophet
    import mlflow

    train_df = pd.read_csv("https://raw.githubusercontent.com/facebook/prophet/main/examples/example_wp_log_peyton_manning.csv")
    # starts on 2007-12-10, ends on 2016-01-20

>     print(f"Training data start: {train_df.ds.iloc[0]}")

Reviewer: Can we remove these print statements?

Author: Removed.

    print(f"Training data end: {train_df.ds.iloc[-1]}")

>     reg_model = Prophet()

Reviewer: Can we show realistic parameter configurations here? i.e., overriding the `changepoint_prior_scale` and perhaps setting `uncertainty_samples` to something like '7' or '14'?

Author: Done.

    reg_model.fit(train_df)

    # Serialize the Model
    with mlflow.start_run():

>         model_info = mlflow.prophet.log_model(reg_model, "prophet-model")

Reviewer: The model fit should be within the MLflow run context.

Author: Fixed.

>         model_info = mlflow.prophet.log_model(reg_model, "prophet-model")

Reviewer: Please extract the parameters from the Prophet model (the ones that are not Objects) and log them to the run.

Author: Done. Note that my change logs the following: 

```python
{'changepoint_prior_scale': 0.5,
 'changepoint_range': 0.8,
 'daily_seasonality': 'auto',
 'growth': 'linear',
 'holidays_prior_scale': 10.0,
 'interval_width': 0.8,
 'logistic_floor': False,
 'mcmc_samples': 0,
 'n_changepoints': 25,
 'seasonality_mode': 'additive',
 'seasonality_prior_scale': 10.0,
 'specified_changepoints': False,
 'uncertainty_samples': 7,
 'weekly_seasonality': 'auto',
 'y_scale': 12.846746888829,
 'yearly_seasonality': 'auto'}
```
and not the ``prophet_model.params`` dict, which is all ndarrays.

> 

Reviewer: Can you add cross validation, performance metrics generation from the CV output and an average value across all backtesting windows to get single values for each of the calculated metrics? 
After generating those metrics, can those then be logged at metrics to the run?

Author: Done.

    # Load saved model
    reg_model_saved = mlflow.pyfunc.load_model(model_info.model_uri)

    # Create a test DataFrame with the "ds" column containing 10 days after the end date in train_df
    test_dates = pd.date_range(start="2016-01-21", end="2016-01-31", freq="D")
    test_df = pd.Series(data=test_dates.values, name="ds").to_frame()

    # Predicted values are in the "yhat" column
    predictions = reg_model_saved.predict(test_df)

>     print(predictions.loc[:, ["ds", "yhat"]].head())

Reviewer: No need to have a print statement in here

Author: Removed.

Output (``Pandas DataFrame``):

> ============  =========

Reviewer: We should include the error estimates output here to show the yhat_lower and yhat_upper values for the prediction intervals. Configuring Prophet's class constructor with the appropriate arg values will enforce the generation of error intervals.

Author: Done. Prophet calculates ``yhat_lower`` and ``yhat_upper`` by default unless we set ``uncertainty_samples`` to 0 which we are setting to 7, so no extra args are necessary.

from Prophet's GitHub repository, containing log number of daily views to Peyton Manning’s Wikipedia page for several years.
A sample of the dataset is as follows:

============= =================
ds            y
============= =================
2007-12-10    9.59076113897809
2007-12-11    8.51959031601596
2007-12-12    8.18367658262066
2007-12-13    8.07246736935477
============= =================

.. code-block:: py

    import pandas as pd
    import numpy as np
    from prophet import Prophet
    import mlflow

    train_df = pd.read_csv("https://raw.githubusercontent.com/facebook/prophet/main/examples/example_wp_log_peyton_manning.csv")
    prophet_model = Prophet(changepoint_prior_scale=0.5, uncertainty_samples=7)

    # Serialize the Model
    with mlflow.start_run():
        prophet_model.fit(train_df)

        # extract and log parameters such as changepoint_prior_scale in the mlflow run
        model_params = {k: v for k, v in vars(prophet_model).items() if np.isscalar(v)}
        mlflow.log_params(model_params)
        model_info = mlflow.prophet.log_model(prophet_model, "prophet-model")

    # Load saved model

>     reg_model_saved = mlflow.pyfunc.load_model(model_info.model_uri)

Reviewer: ```suggestion
    prophet_model_saved = mlflow.pyfunc.load_model(model_info.model_uri)
```

Author: Done.

    # Create a test DataFrame with the "ds" column containing 10 days after the end date in train_df
    test_dates = pd.date_range(start="2016-01-21", end="2016-01-31", freq="D")
    test_df = pd.Series(data=test_dates.values, name="ds").to_frame()

    # Predicted values are in the "yhat" column

>     predictions = reg_model_saved.predict(test_df)

Reviewer: ```suggestion
    predictions = prophet_model_saved.predict(test_df)
```

Author: Done.
